pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/SelectTupleElementTransformFunction.java

```diff
@@ -0,0 +1,100 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import java.util.Comparator;
+import java.util.EnumMap;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.segment.spi.datasource.DataSource;
+import org.apache.pinot.spi.data.FieldSpec;
+
+
+public abstract class SelectTupleElementTransformFunction extends BaseTransformFunction {
+
+  private static final EnumSet<FieldSpec.DataType> SUPPORTED_DATATYPES = EnumSet.of(FieldSpec.DataType.INT,
+      FieldSpec.DataType.LONG, FieldSpec.DataType.FLOAT, FieldSpec.DataType.DOUBLE, FieldSpec.DataType.TIMESTAMP);
+
+  private static final Comparator<FieldSpec.DataType> DOMINANCE_HIERARCHY = Comparator.comparingInt(Enum::ordinal);
+
+  private static final EnumMap<FieldSpec.DataType, EnumSet<FieldSpec.DataType>> ACCEPTABLE_COMBINATIONS =
+      createAcceptableCombinations();
+
+  private final String _name;
+
+  protected List<TransformFunction> _arguments;
+  private TransformResultMetadata _resultMetadata;
+
+  public SelectTupleElementTransformFunction(String name) {
+    _name = name;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    if (arguments.isEmpty()) {
+      throw new IllegalArgumentException(_name + " takes at least one argument");
+    }
+    FieldSpec.DataType dataType = null;
+    for (int i = 0; i < arguments.size(); i++) {
+      TransformFunction argument = arguments.get(i);
+      TransformResultMetadata metadata = argument.getResultMetadata();
+      if (!metadata.isSingleValue()) {
+        throw new IllegalArgumentException(argument.getName() + " at position " + i + " is not single value");
+      }
+      FieldSpec.DataType argumentType = metadata.getDataType();
+      if (!SUPPORTED_DATATYPES.contains(argumentType)) {
+        throw new IllegalArgumentException(argumentType + " not supported. " + SUPPORTED_DATATYPES + " are supported.");
+      }
+      if (dataType == null) {
+        dataType = argumentType;
+      } else if (ACCEPTABLE_COMBINATIONS.get(dataType).contains(argumentType)) {
+        if (DOMINANCE_HIERARCHY.compare(dataType, argumentType) < 0) {
+          dataType = argumentType;
+        }
```

Comment:
We might want an else clause here to reject unacceptable combinations

Reply:
There won't be unacceptable combinations if we get inside this condition. What this logic does is promote the type used for comparison. If the type of the most recent argument dominates the type of the current argument, it is not updated. If the current type dominates the most recent type, it is updated.

Reply:
I've changed this anyway - try to preserve type, when there are mismatches, most of the time promote to DOUBLE, but promote pairs of LONG and TIMESTAMP to TIMESTAMP, pairs of INT and LONG to LONG.

Reply:
outdated code, implemented more laborious promotion algorithm